File: sample-objc/ARGEARSample/UI/UIComponent/Setting/SettingView.m
//
//  SettingView.m
//  ARGEARSample
//
//  Created by Jihye on 10/09/2019.
//  Copyright © 2019 Seerslab. All rights reserved.
//

#import "SettingView.h"

@implementation SettingView

#define MENU_LEADING -243
#define kSettingSegmentedControlBackgroundColor [UIColor colorWithRed:41.f/255.f green:52.f/255.f blue:80.f/255.f alpha:1.0f]

/*
// Only override drawRect: if you perform custom drawing.
// An empty implementation adversely affects performance during animation.
- (void)drawRect:(CGRect)rect {
    // Drawing code
}
*/

- (void)awakeFromNib {
    [super awakeFromNib];

    [_autoSaveLabel setText:[@"auto_save" localized]];
    
    UIView *segmentedControlBackView = [[UIView alloc] initWithFrame:CGRectMake(0, 0, 85, 31)];
    [segmentedControlBackView setBackgroundColor:kSettingSegmentedControlBackgroundColor];
    [[segmentedControlBackView layer] setCornerRadius:15.5f];
    UIImage *clearImage = [self getImage:[UIColor clearColor]];
    [_bitrateSegmentedControl setBackgroundImage:clearImage forState:UIControlStateNormal barMetrics:UIBarMetricsDefault];
    [_bitrateSegmentedControl setTitleTextAttributes:@{NSForegroundColorAttributeName: [UIColor clearColor], NSFontAttributeName: [UIFont systemFontOfSize:13]} forState:UIControlStateNormal];
    [_bitrateSegmentedControl setTitleTextAttributes:@{NSForegroundColorAttributeName: [UIColor clearColor], NSFontAttributeName: [UIFont systemFontOfSize:13]} forState:UIControlStateSelected];
    [_bitrateSegmentedControl setDividerImage:clearImage forLeftSegmentState:UIControlStateNormal rightSegmentState:UIControlStateNormal barMetrics:UIBarMetricsDefault];
    [_bitrateSegmentedControl addSubview:segmentedControlBackView];

    [_bitrateSegmentedControl addSubview:_segmentedControlThumb];
    
    [_appInfoLabel setText:[@"application_info" localized]];
    
    [_appVersionLabel setText:[NSString stringWithFormat:@"v%@", [[[NSBundle mainBundle] infoDictionary] objectForKey:@"CFBundleShortVersionString"]]];
}

- (UIImage *)getImage:(UIColor *)color {
    
    CGRect imageRect = CGRectMake(0.0f, 0.0f, 1.f, 1.f);
    UIGraphicsBeginImageContext(imageRect.size);
    CGContextRef context = UIGraphicsGetCurrentContext();
    CGContextSetFillColorWithColor(context, [color CGColor]);
    CGContextFillRect(context, imageRect);
    UIImage *image = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    return image;
}

- (void)setPreferencesWithAutoSave:(BOOL)autoSave showLandmark:(BOOL)showLandmark videoBitrate:(NSInteger)videoBitrate {
    
    [_autoSaveSwitch setOn:autoSave];
    [_autoSaveSwitch setTag:[_autoSaveSwitch isOn]];
    
    [_faceLandmarkSwitch setOn:showLandmark];
    [_faceLandmarkSwitch setTag:[_faceLandmarkSwitch isOn]];
    
    [_bitrateSegmentedControl setSelectedSegmentIndex:videoBitrate];
    [_bitrateSegmentedControl setTag:videoBitrate];
    [self animateBitrateSegmentedControl:videoBitrate];
}

- (void)open {
    [self setHidden:NO];
    
    __weak SettingView *weakSelf = self;
    [UIView animateWithDuration:0.25 delay:0.0 options:UIViewAnimationOptionCurveEaseInOut animations:^{
        [[weakSelf radiusView] setAlpha:1];
        [[weakSelf menuViewLeading] setConstant:0];

        [self layoutIfNeeded];
    } completion:nil];
}

- (void)close {
    
    __weak SettingView *weakSelf = self;
    [UIView animateWithDuration:0.25 delay:0.0 options:UIViewAnimationOptionCurveEaseInOut animations:^{
        [[weakSelf radiusView] setAlpha:0];
        [[weakSelf menuViewLeading] setConstant:MENU_LEADING];

        [self layoutIfNeeded];
    } completion:^(BOOL finished) {
        [self setHidden:YES];
    }];
}

- (IBAction)handleTapFrom:(UITapGestureRecognizer*)tap {
    if ([tap state] == UIGestureRecognizerStateEnded) {
        
        [self close];
    }
}

- (IBAction)autoSaveSwitchAction:(UISwitch *)sender {
    
    [sender setTag:[sender isOn]];
}

- (IBAction)faceLandmarkSwitchAction:(UISwitch *)sender {
    
    [sender setTag:[sender isOn]];
}

- (IBAction)bitrateSegmentedControlAction:(UISegmentedControl *)sender {
    
    NSInteger selectedSeg = [sender selectedSegmentIndex];
    
    [self animateBitrateSegmentedControl:selectedSeg];

    [sender setTag:selectedSeg];
}

- (void)animateBitrateSegmentedControl:(NSInteger)selectedSeg {

    CGFloat xPosition = 2 + 27 * selectedSeg;
    
    NSArray *segmentTitles = @[@"4M bps",
                               @"2M bps",
                               @"1M bps"
    ];

    __weak SettingView *weakSelf = self;
    [UIView animateWithDuration:0.2f delay:0.0 options:UIViewAnimationOptionCurveEaseInOut animations:^{
        [[weakSelf segmentedControlThumbLeading] setConstant:xPosition];
        [[weakSelf segmentedControlThumbTitleLabel] setText:segmentTitles[selectedSeg]];

        [self layoutIfNeeded];
    } completion:nil];
}

@end

File: sample-swift/ARGearSample/UI/UIComponent/Setting/SettingView.swift
//
//  SettingView.swift
//  ARGearSample
//
//  Created by Jihye on 2020/01/28.
//  Copyright © 2020 Seerslab. All rights reserved.
//

import UIKit

protocol SettingDelegate: class {
    func autoSaveSwitchAction(_ sender: UISwitch)
    func faceLandmarkSwitchAction(_ sender: UISwitch)
    func bitrateSegmentedControlAction(_ sender: UISegmentedControl)
}

class SettingView: UIView {
    
    weak var delegate: SettingDelegate?
    
    @IBOutlet weak var menuView: UIView!
    @IBOutlet weak var touchView: UIView!
    @IBOutlet weak var radiusView: UIView!

    @IBOutlet weak var menuViewLeading: NSLayoutConstraint!   // 243

    @IBOutlet weak var autoSaveSwitch: UISwitch!
    @IBOutlet weak var autoSaveLabel: UILabel!
    
    @IBOutlet weak var faceLandmarkSwitch: UISwitch!
    
    @IBOutlet weak var bitrateSegmentedControl: UISegmentedControl!
    @IBOutlet weak var segmentedControlThumb: UIView!
    @IBOutlet weak var segmentedControlThumbTitleLabel: UILabel!
    @IBOutlet weak var segmentedControlThumbLeading: NSLayoutConstraint!
    
    @IBOutlet weak var appInfoLabel: UILabel!
    @IBOutlet weak var appVersionLabel: UILabel!
    
    let menuLeading: CGFloat = -243
    let kSettingSegmentedControlBackgroundColor: UIColor = UIColor(red: 41.0/255.0, green: 52.0/255.0, blue: 80.0/255.0, alpha: 1.0)

    
    override func awakeFromNib() {
        super.awakeFromNib()
        
        self.autoSaveLabel.text = "auto_save".localized()
        
        let segmentedControlBackView = UIView(frame: CGRect(x: 0, y: 0, width: 85, height: 31))
        segmentedControlBackView.backgroundColor = kSettingSegmentedControlBackgroundColor
        segmentedControlBackView.layer.cornerRadius = 15.5
        
        let clearImage = UIColor.clear.getImage()
        bitrateSegmentedControl.setBackgroundImage(clearImage, for: .normal, barMetrics: .default)
        bitrateSegmentedControl.setTitleTextAttributes([NSAttributedString.Key.foregroundColor: UIColor.clear, NSAttributedString.Key.font: UIFont.systemFont(ofSize: 13)], for: .normal)
        bitrateSegmentedControl.setTitleTextAttributes([NSAttributedString.Key.foregroundColor: UIColor.clear, NSAttributedString.Key.font: UIFont.systemFont(ofSize: 13)], for: .selected)
        bitrateSegmentedControl.setDividerImage(clearImage, forLeftSegmentState: .normal, rightSegmentState: .normal, barMetrics: .default)
        bitrateSegmentedControl.addSubview(segmentedControlBackView)
        bitrateSegmentedControl.addSubview(segmentedControlThumb)

        appInfoLabel.text = "application_info".localized()
        
        if let infoDictionary = Bundle.main.infoDictionary, let versionString = infoDictionary["CFBundleShortVersionString"] as? String {
            appVersionLabel.text = "v\(versionString)"
        }
    }
    
    func setPreferences(autoSave: Bool, showLandmark: Bool, videoBitrate: Int) {
        
        autoSaveSwitch.isOn = autoSave
        faceLandmarkSwitch.isOn = showLandmark
        bitrateSegmentedControl.selectedSegmentIndex = videoBitrate
        
        self.animateBitrateSegmentedControl(selectedIndex: videoBitrate)
    }

    @IBAction func handleTapFrom(_ tap: UITapGestureRecognizer) {
        if tap.state == .ended {
            self.close()
        }
    }
    
    @IBAction func autoSaveSwitchAction(_ sender: UISwitch) {
        guard let delegate = self.delegate else { return }
        
        delegate.autoSaveSwitchAction(sender)
    }
    
    @IBAction func faceLandmarkSwitchAction(_ sender: UISwitch) {
        guard let delegate = self.delegate else { return }

        delegate.faceLandmarkSwitchAction(sender)
    }
    
    @IBAction func bitrateSegmentedControlAction(_ sender: UISegmentedControl) {
        guard let delegate = self.delegate else { return }
        
        delegate.bitrateSegmentedControlAction(sender)

        self.animateBitrateSegmentedControl(selectedIndex: sender.selectedSegmentIndex)
    }
    
    func open() {
        self.isHidden = false
        
        UIView.animate(withDuration: 0.25, delay: 0, options: .curveEaseInOut, animations: {
            self.radiusView.alpha = 1.0
            self.menuViewLeading.constant = 0
            
            self.layoutIfNeeded()
        }, completion: nil)
    }
    
    func close() {
        UIView.animate(withDuration: 0.25, delay: 0, options: .curveEaseInOut, animations: {
            self.radiusView.alpha = 0
            self.menuViewLeading.constant = self.menuLeading
            
            self.layoutIfNeeded()
        }) { finished in
            self.isHidden = true
        }
    }
    
    func animateBitrateSegmentedControl(selectedIndex: Int) {

        let xPosition = CGFloat(2 + 27 * selectedIndex)
        let segmentTitles = ["4M bps",
                             "2M bps",
                             "1M bps"
        ]
        
        UIView.animate(withDuration: 0.2, delay: 0, options: .curveEaseInOut, animations: {
            self.segmentedControlThumbLeading.constant = xPosition
            self.segmentedControlThumbTitleLabel.text = segmentTitles[selectedIndex]
            
            self.layoutIfNeeded()
        }, completion: nil)
    }
}
